druid-shell/src/platform/x11/application.rs

```diff
@@ -81,14 +86,65 @@ impl Application {
             quitting: false,
             windows: HashMap::new(),
         }));
+        let present_opcode = match Application::query_present_opcode(&connection) {
+            Ok(p) => p,
+            Err(e) => {
+                log::info!("failed to find Present extension: {}", e);
+                None
+            }
+        };
+        let xfixes = xcb::query_extension(&connection, "XFIXES")
+            .get_reply()
+            .unwrap();
+        dbg!(xfixes.present(), xfixes.major_opcode());
         Ok(Application {
             connection,
             screen_num,
             window_id,
             state,
+            present_opcode,
         })
     }
 
+    // Check if the Present extension is supported, returning its opcode if it is.
+    fn query_present_opcode(conn: &Rc<Connection>) -> Result<Option<u8>, Error> {
+        let query = xcb::query_extension(&conn, "Present")
+            .get_reply()
+            .context("query Present extension")?;
+
+        if !query.present() {
+            return Ok(None);
+        }
+
+        let opcode = Some(query.major_opcode());
+
+        // If Present is there at all, version 1.0 should be supported. This code
+        // shouldn't have a real effect; it's just a sanity check.
+        let version = xcb::present::query_version(&conn, 1, 0).get_reply()?;
+        log::info!(
+            "X server supports Present version {}.{}",
+            version.major_version(),
+            version.minor_version()
+        );
+
+        // We need the XFIXES extension to use regions. This code looks like it's just doing a
+        // sanity check but it is *necessary*: XFIXES doesn't work until we've done version
+        // negotiation
+        // (https://www.x.org/releases/X11R7.7/doc/fixesproto/fixesproto.txt)
+        let version = xcb::xfixes::query_version(&conn, 5, 0).get_reply()?;
+        log::info!(
+            "X server supports XFIXES version {}.{}",
+            version.major_version(),
+            version.minor_version()
+        );
+
+        Ok(opcode)
+    }
+
+    pub fn present_opcode(&self) -> Option<u8> {
```

Comment:
Let's add `#[inline]` and a doc comment. Also does this need to be `pub`? Let's make it `pub(crate)`.